### ISSUE TITLE
Strange Haxe type inference completion behavior

### ISSUE THREAD

**Issue by fullofcaffeine** [HaxeFoundation/haxe#8326]
From what I've researched so far, this seems to be a Haxe/completion server issue. Please correct me if I'm wrong.

It all started when I converted an Express app from TS to Haxe. I noticed that for the I was not getting the proper type info when I hovered or called `$type()` for the middleware `use` function callbacks params. 

I then started a somewhat long discussion that was summarized [here](https://community.haxe.org/t/type-inference-completion-hint-woes/1754). TL;DR follows.

## TL;DR:

I've created a small project that can be used to reproduce the issue [here](https://github.com/fullofcaffeine/haxe-inference-vshaxe-confusion-). **Check the comments here for specific instructions on how to reproduce it:** https://github.com/fullofcaffeine/haxe-inference-vshaxe-confusion-/blob/master/Main.hx#L9.

The completion UX for these scenarios is terrible. Is that expected / how it's supposed to work? Is there something we can do to improve it? Or is this a bug?

**Comment by Simn** [comment]
This is probably a fanfiction-variant of #7753.

**Comment by Simn** [comment]
This was fixed in https://github.com/HaxeFoundation/haxe/commit/e08e68f698b812778c6c1fda98ffd39a843d24b0
